Improve the documentation for mooc-grader questionnaires (#26)
apluslms/aplus-manual: introduction/07_questionnaires.rst

final course grades.) Setting the difficulty is optional and it can be set even if the 
max points aren't defined in the argument. 

**Other options for the questionnaire** 

The questionnaire directive accepts the following options. These are placed below the directive, 
with the format ``:option: possible value``:

- ``submissions``: max submissions
- ``points-to-pass``: points to pass
- ``feedback``: If set, assumes the defaults for a feedback questionnaire
- ``title``: exercise title
- ``no-override``: If set, the conf.py override setting is ignored
- ``pick_randomly``: integer. The questionnaire selects N questions randomly for the 
  user instead of showing all questions. The random selection changes after the user submits,
  but persists without changes if the user just reloads the web page. (The questionnaire should 
  not include any static text fields between the questions since the text fields are part of 
  the pool from which the questions are randomly selected.)
- ``preserve-questions-between-attempts``: If set, the questions in a ``pick_randomly`` 
  questionnaire are preserved between submission attempts, instead of being resampled 
  after each attempt.

> - ``category``: exercise category

Reviewer: We can add additional information on what a category is. I am not actually sure what a category is 🤔 .

Reviewer 2: Each learning object in A+ belongs to a category. You must have seen the default categories "submit" and "questionnaire" before. Those default names are not very useful, but they can be changed. Teachers can also define the category manually for each exercise if the default is not good.

Reviewer 2: Just to emphasize, categories are not a feature of **questionnaires**, but every exercise has a category. If you don't define any category, then the default value will be used.

  solution after the user has consumed all submission attempts. Can be set true or false. 
  The feedback may reveal the model solution even before the exercise deadline. 
  Note that A+ has a separate feature for showing exercise model solutions after the deadline. 
  The default value can be set in index.rst with the field 
  questionnaire-default-reveal-model-at-max-submissions. By default false.
- ``show-model``: Students may open the model solution in A+ after the module deadline. 
  Can be set to true or false. The default value can be set in index.rst with the field 
  questionnaire-default-show-model. By default true.
- ``allow-assistant-viewing``: Allows assistants to view the submissions of the students. 
  Can be set to true or false. Overrides any options set in the conf.py or config.yaml files.
- ``allow-assistant-grading``: Allows assistants to grade the submissions of the students. 
  Can be set to true or false. Overrides any options set in the conf.py or config.yaml files.
|

Individual questions
--------------------

**Instructions:** Instructions or description for each question can be set below the

> question title (see the examples below).

Reviewer: Add reference to the examples

Reviewer: I will remove the `(see the examples below)` as it is not possible to see how the instructions are set below the question title in the HTML version of the manual, and we do not expect all the user to download the course manual or explore the source code of it. Maybe in the future, we can use annotated directives to explain the questionnaire and some other directives.

> **Hints:** Hints are set below the instructions, possible answer options and correct

Reviewer: A short and clean example on how to use hints may be beneficial, since using hints require to prepend symbols to the answers which are linked to the `value § hint` statement

I also wander what happened with the `%100%` option. Is it still working? @Mankro Do you know? or Maybe @saarasat can give us a quick response 😉

Reviewer 2: `%100%` should still be here. If Saara removed it by accident, then you can add it back.

Reviewer 2: `%100%` has stopped working in A+ v1.8. That must be fixed in this issue: https://github.com/apluslms/a-plus-rst-tools/issues/112

options. The hints are targeted to specific choices and they are shown after answering. 
The format of the hints is ``chosen option or answer § feedback text or hint``. 
The value is the student's submission and it may be prepended with ``!`` in order to 
show the feedback when the student did not answer that value. In freetext questions, 
the value may be prepended with ``regexp``: in order to use regular expressions for matching 
the student's submission.

> **Question type:** The question directives ``pick-one``, ``pick-any``, and ``freetext`` 

Reviewer: The question type should be placed first, since the instructions and hints are basically the directive content of the question type.

start a new question. So for instance ``.. pick-one:: 10`` begins a single-choice question
with the maximum points of 10. 

**Other options for individual questions**

The question directive accepts also the following options:

- ``class``: CSS class
- ``required``: If set, the question is required and empty answers of the questionnaire 
  are rejected and no submission is consumed. 
- ``key``: a manually set key for the question. This affects the HTML input element and the 
  key in the submission data. If no key is set, note that automatically added keys change 
  when the order and amount of questions is modified.

Single-choice and multiple-choice questions
-------------------------------------------

**Correct options**: Correct options in ``pick-one`` (single-choice) and ``pick-any`` (multiple choice)

> directives are marked with ``*``. 

Reviewer: ```suggestion
**Correct answers**: The correct answers for ``pick-one`` (single-choice) and ``pick-any`` (multiple choice)
directives are marked by prefixing  the ``*`` symbol to the corresponding answer/option.
For instance, in the following example the option **b** is marked as the correct answer.

.. code-block: rst

.. pick-one:: 10

  What is 1+1?
    
  a. 1
  *b. 2
  c. 3
```

**Neutral options**: Multiple-choice questions may have neutral options, which are marked 
with ``?``. Neutral options do not affect grading, regardless of whether the student selected 

> them or not. 

Reviewer: Rewording.
```suggestion
**Neutral options**: Multiple-choice questions may have neutral options, which are marked 
by prefixing the ``?`` symbol to the corresponding answer/option. Neutral options do not affect grading, regardless of whether the student selected 
them or not. 
```

> **Initially selected options** Initially selected options may be set with ``+``.

Reviewer: ```suggestion
**Initially selected options** Initially selected options may be set by prefixing the ``+`` symbol to the corresponding  answer/option.
```

The initially selected options are pre-selected when the exercise is loaded. 

> The ``+`` character is written before ``*`` or ``?`` if they are combined.

Reviewer: ```suggestion
The ``+`` character must precede the ``*`` or ``?`` symbols if they are combined, i.e. ``+* OPTION`` or  ``+? OPTION``
```

The ``pick-any`` directive has following options in addition to the common question options:

- ``partial-points``: When set, the question awards points for partially correct submissions. 
  The points scale linearly to the maximum points when more than half of the options 
  are answered correctly.
- ``randomized``: When this option is used, a subset of the answer options (checkboxes) 
  is randomly selected for the user. The random selection changes after the 
  user submits, but persists when the user just reloads the web page. The value of the 
  option is an integer, which is the number of choices to randomly select from all of 
  the defined answer choices in the question. The option correct-count should be also 
  set when this option is used.
- ``correct-count``: The number of correct answer options (checkboxes) to randomly select in the
  randomized pick-any question. This option is used with the randomized option.
- ``preserve-questions-between-attempts``: If set, the answer choices in a randomized 
  question are preserved between submission attempts (instead of being resampled after each attempt).
|

> 

Reviewer: Maybe the Example title is still required here. It may be also really useful to attach the code example before or after the actual example. In the future we can use the tab directive for such a purposes. https://github.com/apluslms/a-plus-rst-tools/pull/83

    !b § If option "an integer" is not chosen, this hint will be shown.
    d § No. This equation has a nice and easy solution.

  .. pick-any:: 10
    :partial-points:

    Checkbox questions defined with ``pick-any`` can have the option ``partial-points``. 
    Students are then granted points also for partially correct answers. You can try it out 
    below.
    
    For instance, in this case there are three correct answers, the grading goes as follows:
    1 correct = 3 points,
    2 correct = 6 points,
    3 correct = 10 points.
    
    And for the wrong answers:
    1 wrong option chosen = 3 points deducted, 
    2 wrong options chosen = 6 points deducted,
    3 wrong options chosen = 10 points deducted.

    When :math:`(x + 1)^2 = 16`, what is :math:`x`?

    a. 4
    *b. an integer
    *c. 3
    d. an irrational number
    e. -3
    *f. -5

Freetext questions
------------------

The ``freetext`` subdirective creates text boxes and grades based on their
input.

The freetext directive also accepts the following options in addition to 
the common question options:

> - ``length``: (horizontal) length for the HTML text input

Reviewer: maybe we should clarify the units we are using. (px, pt, cm, em,  %, ...)

Reviewer 2: Yes, you can check the units. I have a feeling that this option does not accept units and it doesn't even really do anything. Is the length saved in the exercise YAML file at all?

Reviewer: No, the length is not stored in the exercise YAML file. 
![image](https://user-images.githubusercontent.com/10437475/99387704-1d562380-28dd-11eb-83f1-f4f9976a1693.png)
It seems that the length defines the attribute `size` of the input element or the number of columns when the height is defined https://github.com/apluslms/a-plus-rst-tools/blob/94523f890b488a99217049cb9e07577448ce8bda/directives/questionnaire.py#L582, However, I do not see the size attribute or the defined number of cols in the HTML code. 🤔 Maybe, we need to check the questionnaire directive.

Reviewer 2: Yes, the height value changes the text input into multiline textarea element, but length maybe does nothing at all at the moment.

For instance ``freetext:: 5 int``. 

The different comparison methods are ``int``, ``float``, ``string``, 
``subdiff``, ``regexp`` or ``unsortedchars`` (unsorted character set). The ``regexp`` 
compare method takes the correct answer as a regular expression and tries to
match the submission with it. ``subdiff``: works almost like the string method, 
but it can have multiple correct answers separated with ``|`` and if the answer is 
incorrect, it shows the difference of the answer to each correct answer as a hint. 
For example, when the correct answer is *'cat'* and the student answers *'car'*, 
the student receives feedback: *"Correct parts in your answer: ca-."* 

String methods have comparison modifiers that are separated with a hyphen. 
For example, ``.. freetext:: 30 string-ignorews-ignorequotes``. 

The following modifiers are available:

- ``ignorews:`` ignore white space (applies to regexp too)
- ``ignorequotes``: iqnore "quotes" around
- ``requirecase``: require identical lower and upper cases (only with the 
  string and subdiff types)
- ``ignorerepl``: ignore REPL prefixes
- ``ignoreparenthesis``: ignore parenthesis "( )"
|

> 

Reviewer: Example title?. Again, showing both, the example and the code may be beneficial for the reader

  answers of one type and three identical answers of another type.

Additional information
......................

The most recent and complete documentation on this can
be found at `the source code of the A+ RST tools package

> <https://github.com/Aalto-LeTech/a-plus-rst-tools>`_. 

Reviewer 2: The repo was moved to the apluslms organization.

See the source code of `the A+ RST tools questionnaire directive
<https://github.com/Aalto-LeTech/a-plus-rst-tools/blob/master/directives/questionnaire.py>`_
and the corresponding `form implementation in mooc-grader

> <https://github.com/Aalto-LeTech/mooc-grader/blob/master/access/types/forms.py>`_.

Reviewer 2: The repo URL should use the apluslms organization.
